kernel/src/mem.rs

```diff
@@ -1,202 +1,278 @@
-                        .is_some()
-                })
+        }
+    }
+
+    fn mut_map_or<F, R>(&self, default: R, fun: F) -> R
+    where
+        F: FnOnce(&mut [u8]) -> R,
+    {
+        match self.process_id {
+            None => default,
+            Some(pid) => pid.kernel.process_map_or(default, pid, |_| {
+                // Safety: `kernel.process_map_or()` validates that the process still exists
+                // and its memory is still valid. `Process` tracks the "high water mark" of
+                // memory that the process has `allow`ed to the kernel, and will not permit
+                // the process to free any memory after it has been `allow`ed. This guarantees
+                // that the buffer is safe to convert into a slice here.
+                let slice = unsafe { slice::from_raw_parts_mut(self.ptr, self.len) };
+                fun(slice)
+            }),
+        }
+    }
+}
+
+impl Read for ReadWriteAppSlice {
+    fn len(&self) -> usize {
+        self.process_id
+            .map_or(0, |pid| pid.kernel.process_map_or(0, pid, |_| self.len))
+    }
+
+    fn ptr(&self) -> *const u8 {
+        if self.len == 0 {
+            0x0 as *const u8
         } else {
-            false
+            self.ptr
         }
     }
 
-    /// Returns an iterator over the slice
-    ///
-    /// See
-    /// [`std::slice::iter()`](https://doc.rust-lang.org/std/primitive.slice.html#method.iter).
-    ///
-    /// Internally this uses
-    /// [`AsRef`](struct.AppSlice.html#impl-AsRef<[T]>), hence when
-    /// the app dies, restarts or the
-    /// [`AppId`](crate::callback::AppId) changes for any other
-    /// reason, the iterator will be of zero length.
-    pub fn iter(&self) -> slice::Iter<T> {
-        self.as_ref().iter()
+    fn map_or<F, R>(&self, default: R, fun: F) -> R
+    where
+        F: FnOnce(&[u8]) -> R,
+    {
+        match self.process_id {
+            None => default,
+            Some(pid) => pid.kernel.process_map_or(default, pid, |_| {
+                let slice = unsafe { slice::from_raw_parts(self.ptr, self.len) };
+                fun(slice)
+            }),
+        }
     }
+}
 
-    /// Returns an iterator that allows modifying each value
-    ///
-    /// See
-    /// [`std::slice::iter_mut()`](https://doc.rust-lang.org/std/primitive.slice.html#method.iter_mut).
-    ///
-    /// Internally this uses
-    /// [`AsMut`](struct.AppSlice.html#impl-AsMut<[T]>), hence when
-    /// the app dies, restarts or the
-    /// [`AppId`](crate::callback::AppId) changes for any other
-    /// reason, the iterator will be of zero length.
-    pub fn iter_mut(&mut self) -> slice::IterMut<T> {
-        self.as_mut().iter_mut()
+/// Read-only memory region of a process, shared with the kernel
+pub struct ReadOnlyAppSlice {
+    ptr: *const u8,
+    len: usize,
+    process_id: Option<AppId>,
+}
+
+impl ReadOnlyAppSlice {
+    pub(crate) unsafe fn new(ptr: *const u8, len: usize, process_id: AppId) -> Self {
```

Comment:
This needs safety comment / documentation (and pub interface documentation)

Reply:
This should be fixed in #2464.